{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2814ac3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "###!/usr/bin/env \n",
    "# -*- coding: utf-8 -*-\n",
    "\"\"\"\n",
    "Created on Fri Nov 16 14:35:31 2018\n",
    "First Version was written on Tue Aug  9 08:40:55 2016\n",
    "\n",
    "@author: alexanderkutschera\n",
    "\"\"\"\n",
    "# Script edited by me and chatGPT 13/05/2025, changes:\n",
    "## Converted from python2 to python3\n",
    "## Removed soon to be deprecated Bio.Applications dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "016b126a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import csv\n",
    "from Bio import SeqIO\n",
    "import subprocess\n",
    "from Bio.Blast import NCBIXML\n",
    "\n",
    "# Set paths\n",
    "data_path = \"/home/callum/rotation2/genome_comparison/\"\n",
    "proteomes_path = os.path.join(data_path, \"proteome_database\")\n",
    "query_path = os.path.join(data_path, \"query_PAO1\")\n",
    "blastp_path = \"/home/callum/blast/ncbi-blast-2.16.0+/bin/blastp\"\n",
    "\n",
    "# Gene lists\n",
    "raetz_genes = [\"\", \"LpxA\", \"LpxC\", \"LpxD\", \"LpxH\", \"LpxB\", \"LpxK\", \"WaaA\", \"PSPTO_0182\", \"HtrB\"]\n",
    "core_genes = [\"\", \"MsbA\", \"WaaL\", \"WapR\", \"putative_glycosyl_transferase\", \"DnpA\", \"mig-14\", \"WapH\", \"putative_carbamoyltransferase\", \"putative_kinase\", \"WapQ\", \"WaaP\", \"WaaG\", \"WaaC\", \"WaaF\"]\n",
    "osa_genes = [\"\", \"Wzz2\", \"Wzz\", \"WbpA\", \"WbpB\", \"WbpC\", \"WbpD\", \"WbpE\", \"Wzy\", \"Wzx\", \"HisH2\", \"HisF2\", \"WbpG\", \"WbpH\", \"WbpI\", \"WbpJ\", \"WbpK\", \"WbpL\", \"WbpM\"]\n",
    "cpa_genes = [\"\", \"WbpZ\", \"WbpY\", \"WbpX\", \"Wzt\", \"Wzm\", \"WbpW\", \"Gmd\", \"Rmd\", \"PA5455\", \"PA5456\", \"PA5457\", \"PA5458\", \"PA5459\"]\n",
    "mod_genes = [\"\", \"ArnT\", \"EptA\", \"PagL\", \"LpxO\", \"LpxF\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6ddef27",
   "metadata": {},
   "outputs": [],
   "source": [
    "# All genes\n",
    "all_genes = [gene[:-6] for gene in os.listdir(query_path) if gene.endswith(\".fasta\")]\n",
    "\n",
    "# Choose gene list\n",
    "name = \"biosynthesis\"\n",
    "genelist = all_genes\n",
    "\n",
    "#use if you want to blast specific proteomes\n",
    "#proteomes = [\"Pci_JBC1_proteome\", \"Pst_DC3000_proteome\"] \n",
    "\n",
    "# Get list of proteomes\n",
    "proteomes = [filename[:-4] for filename in os.listdir(proteomes_path) if filename.endswith(\".faa\")]\n",
    "print(f\"You have {len(proteomes)} proteomes in your list\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75a2f0e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---- blast ----\n",
    "\n",
    "# Initialize dictionaries\n",
    "d = {name: genelist}\n",
    "n = {name: genelist}\n",
    "\n",
    "# Run BLAST\n",
    "for count, proteome in enumerate(proteomes, start=1):\n",
    "    # Intialise lists\n",
    "    percentage_list = []\n",
    "    genename_list = []\n",
    "\n",
    "    # Iterate through list of genes\n",
    "    for gene in genelist[1:]:\n",
    "        print(f\"BLASTing {gene} on {proteome} ...\")\n",
    "\n",
    "        # Get query sequence\n",
    "        qseq = next(SeqIO.parse(os.path.join(query_path, gene + \".fasta\"), \"fasta\"))\n",
    "        print(f\"Query: {len(qseq.seq)}\")\n",
    "\n",
    "        # Now depends on subprocess\n",
    "        query_file = os.path.join(query_path, gene) + \".fasta\"\n",
    "        db_path = os.path.join(proteomes_path, proteome)\n",
    "        out_file = os.path.join(data_path, \"result.xml\")\n",
    "\n",
    "        cmd = [\n",
    "            blastp_path,\n",
    "            \"-query\", query_file,\n",
    "            \"-db\", db_path,\n",
    "            \"-evalue\", \"1\",\n",
    "            \"-outfmt\", \"5\",  # XML format\n",
    "            \"-out\", out_file\n",
    "        ]\n",
    "        result = subprocess.run(cmd, capture_output=True, text=True)\n",
    "\n",
    "        if result.returncode != 0:\n",
    "            print(f\"Error running BLAST: {result.stderr}\")\n",
    "        else:\n",
    "            print(\"BLAST run completed.\")\n",
    "\n",
    "        # Check the results\n",
    "        with open(os.path.join(data_path, \"result.xml\")) as result_handle:\n",
    "            blast_record = NCBIXML.read(result_handle)\n",
    "        # If there are no alignments, there are no hits\n",
    "        if len(blast_record.alignments) == 0:\n",
    "            percentage = 0\n",
    "            genename = \"no hit\"\n",
    "        # Otherwise this code just records the first hit by default\n",
    "        else:\n",
    "            length = blast_record.alignments[0].length\n",
    "            title = blast_record.alignments[0].title\n",
    "            # I don't understand what the PE bit is for\n",
    "            if \"PE=\" not in title:\n",
    "                genename = title[title.find(\"GN=\")+3:title.find(\"SV=\")]\n",
    "            else:\n",
    "                genename = title[title.find(\"GN=\")+3:title.find(\"PE=\")]\n",
    "\n",
    "            hsp = blast_record.alignments[0].hsps[0]\n",
    "            identity = hsp.identities\n",
    "            print(f\"identities: {hsp.identities}\")\n",
    "            print(f\"length: {length}\")\n",
    "            print(f\"alignlength: {hsp.align_length}\")\n",
    "            percentage = (float(identity) / float(length)) * 100\n",
    "        # Get save gene and percentage similarity\n",
    "        print(f\"Identity percentage is {percentage:.2f} %\\n\")\n",
    "        percentage_list.append(percentage)\n",
    "        genename_list.append(genename)\n",
    "\n",
    "    percentage_list.insert(0, proteome)\n",
    "    genename_list.insert(0, proteome)\n",
    "    d[proteome] = percentage_list\n",
    "    n[proteome] = genename_list\n",
    "\n",
    "# Save results\n",
    "with open(os.path.join(data_path, \"final_results.csv\"), \"w\", newline='') as csv_file:\n",
    "    writer = csv.writer(csv_file)\n",
    "    writer.writerow(d[name])\n",
    "    for proteome in proteomes:\n",
    "        writer.writerow(d[proteome])\n",
    "\n",
    "with open(os.path.join(data_path, \"final_results_genenames.csv\"), \"w\", newline='') as csv_file:\n",
    "    writer = csv.writer(csv_file)\n",
    "    writer.writerow(n[name])\n",
    "    for proteome in proteomes:\n",
    "        writer.writerow(n[proteome])\n",
    "\n",
    "print(\"\\ndone!\")\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
